### PR TITLE
LPS-186246 | Master

### PR DESCRIPTION
**Pull request by Natanaelbenvindo**
https://issues.liferay.com/browse/LPS-186246
https://issues.liferay.com/browse/LPS-186247
https://issues.liferay.com/browse/LPS-186248

### CODE REVIEW

**Comment by Natanaelbenvindo** [comment]
ci:test:sf

File: modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+			ObjectAdmin.openObjectAdmin();
+
+			ObjectPortlet.selectCustomObject(label = "Custom Object 186246");
+
+			ObjectAdmin.goToFieldsTab();
+
+			ObjectAdmin.addObjectFieldViaUI(
+				fieldLabel = "Custom Date Time",
+				fieldTimeStorage = "Convert to UTC",
+				fieldType = "Date and Time");
```
Comment (andreefarias) [inline modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase:46]:
```
ObjectAdmin.addObjectFieldViaUI(
	fieldLabel = "Custom Date Time",
	fieldTimeStorage = "fieldTimeStorage = "Use Input as Entered",
	fieldType = "Date and Time");
```

Comment (andreefarias) [reply modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase:46]: Use Input as Entered here:

```
ObjectAdmin.addObjectFieldViaUI(
	fieldLabel = "Custom Date Time",
	fieldTimeStorage = "fieldTimeStorage = "Use Input as Entered",
	fieldType = "Date and Time");
```

File: modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase [hunk 2]
```diff
@@ -64,5 +64,95 @@ definition {
...
+			JSONUser.addUser(
+				userEmailAddress = "userea@liferay.com",
+				userFirstName = "userfn",
+				userLastName = "userln",
+				userScreenName = "usersn");
+
+			JSONUser.setFirstPassword(
+				agreeToTermsAndAnswerReminderQuery = "true",
+				requireReset = "false",
+				userEmailAddress = "userea@liferay.com");
```
Comment (andreefarias) [inline modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase:123]: This step isn't necessary.

File: modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+				fieldType = "Date and Time");
+
+			ObjectAdmin.publishObjectViaAPI(objectName = "CustomObject186246");
+		}
+
+		task ("When the user adds an object entry") {
+			ObjectAdmin.addObjectSingleFieldEntryViaAPI(
+				fieldName = "customDateTime",
+				objectName = "CustomObject186246",
+				value = "2023-06-01T12:00:00.000Z");
```
Comment (andreefarias) [inline modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase:55]: Could use a date with AM?

**Comment by andreefarias** [comment]
ci:test:sf

File: modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase (hunk 2 above)
Comment (Natanaelbenvindo) [reply modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase]: ok

**Review by andreefarias — APPROVED** [review]
Nice! Good job @Natanaelbenvindo!

LGTM! 🎉

File: modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+			PortalInstances.tearDownCP();
+		}
+	}
+
+	@description = "LPS-143068 - Verify if is possible to add DateTime on entries from Custom Object"
+	@priority = 5
+	test CanAddEntry {
+		property portal.acceptance = "true";
+
+		task ("Given a custom object with a date and time field as input as entered") {
```
Comment (timpak) [inline modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase:29]: Given a custom object with a date and time field and the time storage set to user input

File: modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+			ObjectAdmin.addObjectViaAPI(
+				labelName = "Custom Object 186246",
+				objectName = "CustomObject186246",
+				pluralLabelName = "Custom Objects 186246");
+
+			ObjectAdmin.publishObjectViaAPI(objectName = "CustomObject186246");
```
Comment (timpak) [inline modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase:35]: We shouldn't be publishing the object definition until we're done setting it up

File: modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+			ObjectAdmin.openObjectAdmin();
+
+			ObjectPortlet.selectCustomObject(label = "Custom Object 186246");
+
+			ObjectAdmin.goToFieldsTab();
+
+			ObjectAdmin.addObjectFieldViaUI(
+				fieldLabel = "Custom Date Time",
+				fieldTimeStorage = "Use Input as Entered",
+				fieldType = "Date and Time");
```
Comment (timpak) [inline modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase:46]: Can we add the field using our API instead?

File: modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+		task ("When the user adds an object entry") {
+			ObjectAdmin.addObjectSingleFieldEntryViaAPI(
+				fieldName = "customDateTime",
+				objectName = "CustomObject186246",
+				value = "2023-06-01T09:00:00.000Z");
+
+			ObjectAdmin.goToCustomObject(objectName = "CustomObject186246");
+		}
+
+		task ("Then the date and time is seen on the table entries") {
```
Comment (timpak) [inline modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase:60]: Then the date and time seen on the table

File: modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase [hunk 8]
```diff
@@ -64,5 +64,90 @@ definition {
 		}
 	}
 
+	@description = "LPS-143068  - Verify if is possible to see DateTime with differents Time Zone"
```
Comment (timpak) [inline modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase:67]: Verify the date and time are adjusted to the user's time zone

File: modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase [hunk 9]
```diff
@@ -64,5 +64,90 @@ definition {
...
+	@priority = 5
+	test CanSeeDifferentTimeZone {
+		property portal.acceptance = "true";
+
+		task ("Given a custom object with a date and time field as UTC") {
```
Comment (timpak) [inline modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase:72]: Given a custom object with a date and time field as UTC and an object entry

File: modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase [hunk 10]
```diff
@@ -64,5 +64,90 @@ definition {
...
+			ObjectAdmin.addObjectViaAPI(
+				labelName = "Custom Object 186248",
+				objectName = "CustomObject186248",
+				pluralLabelName = "Custom Objects 186248");
+
+			ObjectAdmin.publishObjectViaAPI(objectName = "CustomObject186248");
```
Comment (timpak) [inline modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase:78]: We shouldn't be publishing here

File: modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase [hunk 11]
```diff
@@ -64,5 +64,90 @@ definition {
...
+			ObjectAdmin.openObjectAdmin();
+
+			ObjectPortlet.selectCustomObject(label = "Custom Object 186248");
+
+			ObjectAdmin.goToFieldsTab();
+
+			ObjectAdmin.addObjectFieldViaUI(
+				fieldLabel = "Custom Date Time",
+				fieldTimeStorage = "Convert to UTC",
+				fieldType = "Date and Time");
```
Comment (timpak) [inline modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase:89]: Same thing as the previous test, can we add the field using the API instead?

File: modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase [hunk 12]
```diff
@@ -64,5 +64,90 @@ definition {
...
+
+			UserNavigator.gotoUser(userScreenName = "test");
+
+			UserNavigator.gotoDisplaySettings();
+
+			Select(
+				locator1 = "Select#TIME_ZONE",
+				value1 = "(UTC) Coordinated Universal Time");
+
+			PortletEntry.save();
```
Comment (timpak) [inline modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase:103]: I'm not sure what this is accomplishing in terms of the test, can you explain why were adding these steps?

File: modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase [hunk 13]
```diff
@@ -64,5 +64,90 @@ definition {
...
+		}
+
+		task ("And given an object entry") {
+			ObjectAdmin.addObjectSingleFieldEntryViaAPI(
+				fieldName = "customDateTime",
+				objectName = "CustomObject186248",
+				value = "2023-06-01T12:00:00.000Z");
+		}
```
Comment (timpak) [inline modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase:111]: This should be part of the previous given

File: modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase [hunk 14]
```diff
@@ -64,5 +64,90 @@ definition {
...
+				userFirstName = "userfn",
+				userLastName = "userln",
+				userScreenName = "usersn");
+
+			JSONRole.assignRoleToUser(
+				roleTitle = "Administrator",
+				userEmailAddress = "userea@liferay.com");
+		}
+
+		task ("When access the user") {
```
Comment (timpak) [inline modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase:125]: This isn't following the original test design

File: modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase [hunk 15]
```diff
@@ -64,5 +64,90 @@ definition {
...
+
+			UserNavigator.gotoUser(userScreenName = "usersn");
+
+			UserNavigator.gotoDisplaySettings();
+
+			Select(
+				locator1 = "Select#TIME_ZONE",
+				value1 = "(UTC -03:00) Brasilia Time");
+
+			PortletEntry.save();
```
Comment (timpak) [inline modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase:140]: Do we need to do this as `usersn` or can we set the the time zone for the user as `Test Test`?

File: modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase [hunk 16]
```diff
@@ -64,5 +64,90 @@ definition {
...
+
+			ObjectAdmin.goToCustomObject(objectName = "CustomObject186248");
+		}
+
+		task ("Then assert that time zone is correct") {
```
Comment (timpak) [inline modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase:145]: Then the time is adjusted to the user's time zone

File: modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase [hunk 17]
```diff
@@ -149,5 +149,59 @@ definition {
 		}
 	}
 
+	@description = "LPS-143068 - Verify if is possible to edit DateTime on entries from Custom Object"
```
Comment (timpak) [inline modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase:152]: Verify a custom object entry with a date and time field can be updated

File: modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase [hunk 18]
```diff
@@ -149,5 +149,59 @@ definition {
...
+	@priority = 5
+	test CanUpdateEntry {
+		property portal.acceptance = "true";
+
+		task ("Given a custom object with a date and time field as UTC") {
```
Comment (timpak) [inline modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase:157]: Given a custom object with a date and time field and an object entry

File: modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase [hunk 19]
```diff
@@ -149,5 +149,59 @@ definition {
...
+			ObjectAdmin.addObjectViaAPI(
+				labelName = "Custom Object 186247",
+				objectName = "CustomObject186247",
+				pluralLabelName = "Custom Objects 186247");
+
+			ObjectAdmin.publishObjectViaAPI(objectName = "CustomObject186247");
```
Comment (timpak) [inline modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase:163]: Same thing here, we shouldn't be publishing the object until we're done setting it up

File: modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase [hunk 20]
```diff
@@ -149,5 +149,59 @@ definition {
...
+			ObjectAdmin.openObjectAdmin();
+
+			ObjectPortlet.selectCustomObject(label = "Custom Object 186247");
+
+			ObjectAdmin.goToFieldsTab();
+
+			ObjectAdmin.addObjectFieldViaUI(
+				fieldLabel = "Custom Date Time",
+				fieldTimeStorage = "Convert to UTC",
+				fieldType = "Date and Time");
```
Comment (timpak) [inline modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase:174]: Same thing here, can we use our API instead?

File: modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase [hunk 21]
```diff
@@ -149,5 +149,59 @@ definition {
...
+		task ("And given an object entry added by user") {
+			ObjectAdmin.addObjectSingleFieldEntryViaAPI(
+				fieldName = "customDateTime",
+				objectName = "CustomObject186247",
+				value = "2023-06-01T12:00:00.000Z");
+
+			ObjectAdmin.goToCustomObject(objectName = "CustomObject186247");
+		}
+
+		task ("When the user update the entry") {
```
Comment (timpak) [inline modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase:188]: `When the user updates the entry`

Please pay careful attention to the wording on the spreadsheet

File: modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase [hunk 22]
```diff
@@ -149,5 +149,59 @@ definition {
...
+			Type(
+				locator1 = "ObjectPortlet#VIEW_EMPTY_LABEL_IN_DATE_ENTRY",
+				value1 = "07/02/2024 10:00 PM");
+
+			PortletEntry.save();
+
+			Navigator.gotoBack();
+		}
+
+		task ("Then the date and time is updated on the table entries") {
```
Comment (timpak) [inline modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase:200]: Then the date and time are updated on the table

**Review by timpak — CHANGES_REQUESTED** [review]
@Natanaelbenvindo please see my feedback below. Some notes:
- Do not deviate from the steps on the spreadsheet after they've been reviewed. If you need to modify the steps, I need to review them again. Pay careful attention to the wording and follow them exactly. Adding typos and grammar mistakes creates more work for the both of us.
- Do not modify the wording of the description and steps at all. If they need to be modified, it must pass through me first. 
- Please familiarize yourself with the user flow for objects. We should be adding the field before publishing the object. We shouldn't be able to publish the object definition without adding a field.

Modifying the steps and wording without clearing it with me generates a lot of extra work on my end during PR review. In the future, please do not do this without talking to me first.

@andreefarias please flag issues like during the initial review and push the PR back to the author to follow the original test design. We shouldn't be deviating from it unless we have a very good reason.

**Comment by andreefarias** [comment]
Hey @timpak,

@Natanaelbenvindo wrote the description/task looking at the ticket but forgot to update the ticket after you updated the sheets.

My review was based on the ticket description/task. Some technical errors were not caught by me, it won't happen again. Thanks for bringing it to me.

**Comment by timpak** [comment]
@Natanaelbenvindo if this is ready for review, please remove the `Author Action Required` label, remove yourself as an assignee. and assign the PR to me. Thanks!

File: modules/apps/object/object-test/src/testFunctional/macros/json/JSONObject.macro [hunk 23]
```diff
@@ -385,6 +412,35 @@ definition {
...
+						"value": "${timeStorageOption}"
+						}
+						
+						],
+						"required": ${isRequired}
+					}'
+			''';	
+		}
+		
+
```
Comment (timpak) [inline modules/apps/object/object-test/src/testFunctional/macros/json/JSONObject.macro:443]: It looks like you're adding this twice

File: modules/apps/object/object-test/src/testFunctional/paths/ObjectField.path [hunk 24]
```diff
@@ -137,6 +137,16 @@
 		<td>//button[@class='dropdown-item']//div[(text()="${key_fieldType}")]</td>
 		<td></td>
 	</tr>
+	<tr>
+		<td>TIME_STORAGE_OPTION</td>
+		<td>//button[contains(@class,'dropdown-item')]//div[(text()="${key_fieldTimeStorage}")]</td>
+		<td></td>
+	</tr>
```
Comment (timpak) [inline modules/apps/object/object-test/src/testFunctional/paths/ObjectField.path:144]: Out of curiosity, can we use an existing path for this dropdown menu option? We should a generic path for this

File: modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase [hunk 25]
```diff
@@ -0,0 +1,64 @@
...
+				pluralLabelName = "Custom Objects 186246");
+
+			ObjectAdmin.addObjectFieldViaAPI(
+				fieldBusinessType = "DateTime",
+				fieldLabelName = "Custom Date Time",
+				fieldName = "customDateTime",
+				fieldType = "DateTime",
+				isRequired = "false",
+				objectName = "CustomObject186246",
+				timeStorageOption = "useInputAsEntered");
```
Comment (timpak) [inline modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase:42]: Can we add the field using the UI here? We should have at least one test that does this

File: modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase [hunk 26]
```diff
@@ -0,0 +1,64 @@
...
+
+			ObjectAdmin.publishObjectViaAPI(objectName = "CustomObject186246");
+		}
+
+		task ("When the user adds an object entry") {
+			ObjectAdmin.addObjectSingleFieldEntryViaAPI(
+				fieldName = "customDateTime",
+				objectName = "CustomObject186246",
+				value = "2023-06-01T09:00:00.000Z");
```
Comment (timpak) [inline modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase:51]: Can we add the entry using the UI here? We should have at least one test that does this

File: modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase [hunk 27]
```diff
@@ -60,5 +60,74 @@ definition {
...
+
+			UserNavigator.gotoDisplaySettings();
+
+			Select(
+				locator1 = "Select#TIME_ZONE",
+				value1 = "(UTC -03:00) Brasilia Time");
+
+			PortletEntry.save();
+
+			ObjectAdmin.goToCustomObject(objectName = "CustomObject186248");
```
Comment (timpak) [inline modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase:114]: We don't need this step here

File: modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase [hunk 28]
```diff
@@ -60,5 +60,74 @@ definition {
...
+
+			User.logoutAndLoginPG(
+				userLoginEmailAddress = "userea@liferay.com",
+				userLoginFullName = "userfn userln");
+		}
+
+		task ("AWhen the user views the entry") {
```
Comment (timpak) [inline modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase:121]: `When the user views the entry`

Please be careful with typos

File: modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase [hunk 29]
```diff
@@ -129,5 +129,53 @@ definition {
...
+				timeStorageOption = "convertToUTC");
+
+			ObjectAdmin.publishObjectViaAPI(objectName = "CustomObject186247");
+
+			ObjectAdmin.addObjectSingleFieldEntryViaAPI(
+				fieldName = "customDateTime",
+				objectName = "CustomObject186247",
+				value = "2023-06-01T12:00:00.000Z");
+
+			ObjectAdmin.goToCustomObject(objectName = "CustomObject186247");
```
Comment (timpak) [inline modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase:159]: This should be part of the `When`, it's not part of the setup to create the object and the entry

File: modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase [hunk 30]
```diff
@@ -129,5 +129,53 @@ definition {
...
+		}
+
+		task ("When the user update the entry") {
```
Comment (timpak) [inline modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsDateAndTime.testcase:162]: `When the user updates the entry`

Please pay attention to the wording. This is the second typo I found.

**Review by timpak — CHANGES_REQUESTED** [review]
Hey @Natanaelbenvindo please see my comments below. Also, can you fix the typo in your third commit message? You didn't capitalize `add`. Please pay attention to details, especially if they were corrected originally. 

Sometimes, it's easier to copy+paste text instead of typing it out. That way, any mistakes would be on the person who originally wrote the text.

**Review by timpak — APPROVED** [review]

**Comment by timpak** [comment]
ci:forward

**Pull request closed without merging**